src/components/PaginatedVirtualList/usePaginatedVirtualList.ts

```diff
@@ -0,0 +1,152 @@
+    totalItemCount: rowCount,
+    currentPage,
+  });
+
+  const scrollToNextPage = useCallback(() => {
+    const nextPage = currentPage + 1;
+    if (nextPage < totalPageCount) {
+      setCurrentPage(nextPage);
+    }
+  }, [currentPage, totalPageCount]);
+
+  const scrollToPrevPage = useCallback(() => {
+    const nextPage = currentPage - 1;
+    if (currentPage !== 0) {
+      setCurrentPage(nextPage);
+    }
+  }, [currentPage]);
+
+  // The following useEffect is necessary to scroll to the correct position when the user
+  // scrolls to the next or previous page.
+  useEffect(() => {
+    if (prevPage === undefined || isScrollerScroll.current) {
+      isScrollerScroll.current = false;
+      return;
+    }
+    if (prevPage < currentPage) {
+      // If we're scrolling to the next page, we want to scroll to the top of the next page
+      ref.current?.scrollToIndex({
+        index: paginationOffset,
+        align: "end",
+      });
+      leaveBreadcrumb(
+        "PaginatedVirtualList",
+        {
+          message: "scrollToNextPage",
+          paginationOffset,
+          paginationThreshold,
+          currentPage,
+          totalPageCount,
+        },
+        "process"
+      );
+    } else {
+      // This setTimeout is necessary because the first scroll doesn't always work
+      // I'm not sure why, but this seems to fix it ¯\_(ツ)_/¯
+      setTimeout(() => {
+        ref.current?.scrollToIndex({
+          index: pageSize - paginationOffset,
+          align: "start",
+        });
+      });
+      leaveBreadcrumb(
+        "PaginatedVirtualList",
+        {
+          message: "scrollToPrevPage",
+          offsetCompensation,
+          paginationOffset,
+          paginationThreshold,
+          currentPage,
+          totalPageCount,
+        },
+        "process"
+      );
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [currentPage]);
+
+  // startingIndex is the index of the first item in the list with respect to the visible page
+  const startingIndex = currentPage * paginationThreshold - offsetCompensation;
+
+  const scrollToLine = useCallback(
+    (index: number) => {
+      const nextPage = Math.floor(index / paginationThreshold);
+      if (nextPage !== currentPage) {
+        setCurrentPage(nextPage);
+        isScrollerScroll.current = true;
+      }
+      const nextScrollIndex =
+        index -
+        calculateStartingIndex({
+          page: nextPage,
+          paginationThreshold,
+          paginationOffset,
+        });
+
+      leaveBreadcrumb(
+        "PaginatedVirtualList",
+        {
+          message: "scrollToIndex",
+          index,
+          nextScrollIndex,
+          nextPage,
+          currentPage,
+          paginationThreshold,
+          paginationOffset,
+          startingIndex,
+        },
+        "process"
+      );
+      ref.current?.scrollToIndex({
+        index: nextScrollIndex,
+        align: "start",
+      });
```

Comment:
does this still need a timeout to accommodate for the race between updating the current page and scrolling?